File: client/request_handler.cpp
#include "request_handler.h"
#include "protocol.h"
#include "file_utils.h"
#include "AESWrapper.h"
#include "RSAWrapper.h"
#include "cksum.h"
#include <boost/asio.hpp>
#include <fstream>
#include <iostream>
#include <filesystem>

#define NUM_OF_TRIES 3
#define RSA_KEY_BITS_SIZE 1024
#define PACKET_SIZE 1024
#define AES_KEY_SIZE 32

RequestHandler::RequestHandler()
	: connection_handler(), username(), filename(), client_id(), private_rsa_key(), aes_key()
{
    username = read_username() + '\0';
    file_path = read_file_path();
    std::filesystem::path path(file_path);
    filename = path.filename().string();
}

RequestHandler::~RequestHandler()
{
	// destructor
}

void RequestHandler::register_user()
{
    // handle the registration request
    Request_t request;
    request.code = REGISTER;
    request.append_to_payload(username, NAME_SIZE);   

    Response_t response = connection_handler.exchange_messages(request);
    if (response.code == REGISTRATION_SUCCESSFUL)
    {
        client_id = response.read_from_payload(CLIENT_ID_OFFSET, CLIENT_ID_SIZE);
        write_username(username);
        write_client_id(client_id);
		std::cout << "User " << username << " registered successfully" << std::endl;
    }
    if (response.code == REGISTRATION_FAILED)
    {
        throw std::runtime_error("Registration failed");
    }
    if (response.code == GENERAL_ERROR)
    {
        throw std::runtime_error("General server error");
    }
}

void RequestHandler::login()
{
    // handle the login request
    Request_t request;
	client_id = read_client_id();
    request.assign_client_id(client_id);
    request.code = LOGIN;
    request.append_to_payload(username, NAME_SIZE);

    Response_t response = connection_handler.exchange_messages(request);
    if (response.code == LOGIN_SUCCESSFUL)
    {
        std::string rsa_key = read_rsa_key();
        aes_key = response.read_from_payload(AES_KEY_OFFSET, response.payload_size - AES_KEY_OFFSET);
        RSAPrivateWrapper rsa_private_wrapper = RSAPrivateWrapper(rsa_key);
        aes_key = rsa_private_wrapper.decrypt(aes_key);
		std::cout << "User " << username << " logged in successfully" << std::endl;
    }
    if (response.code == LOGIN_FAILED)
    {
		std::filesystem::remove("me.info");
		std::filesystem::remove("priv.key");
		std::cout << "User " << username << " not found, registering..." << std::endl;
        register_user();
        exchange_keys();
    }
    if (response.code == GENERAL_ERROR)
    {
        throw std::runtime_error("General server error");
    }
}

void RequestHandler::exchange_keys()
{
    // Generate RSA keys
    RSAPrivateWrapper rsa_private_wrapper = RSAPrivateWrapper();
    std::string rsa_private_key = rsa_private_wrapper.getPrivateKey();
    std::string rsa_public_key = rsa_private_wrapper.getPublicKey();
    write_rsa_private_key(rsa_private_key);

    // Handle the exchange keys request
    Request_t request;
    request.assign_client_id(client_id);
    request.code = SEND_PUBLIC_KEY;
    request.append_to_payload(username, NAME_SIZE);
    request.append_to_payload(rsa_public_key, PUBLIC_KEY_SIZE);

    Response_t response = connection_handler.exchange_messages(request);
    if (response.code == AES_KEY_EXCHANGE)
    {
        aes_key = response.read_from_payload(AES_KEY_OFFSET, response.payload_size - AES_KEY_OFFSET);
        aes_key = rsa_private_wrapper.decrypt(aes_key);
		std::cout << "AES key exchanged successfully" << std::endl;
    }
    if (response.code == GENERAL_ERROR)
    {
        throw std::runtime_error("General server error");
    }
}

void RequestHandler::backup_file()
{
    // opening the file
	std::ifstream original_file(file_path, std::ios::binary);
    if (!original_file.is_open())
    {
        throw std::runtime_error("File not found");
    }

    //computing the original file checksum
    std::string cksum_string = compute_file_crc(file_path);
    unsigned long int cksum = std::stoul(cksum_string.substr(0, cksum_string.find('\t')));

	// encrypting the file
	AESWrapper aes_wrapper = AESWrapper(reinterpret_cast<const unsigned char*>(aes_key.c_str()), AES_KEY_SIZE);
    std::string encrypted_filename = aes_wrapper.encrypt_file(file_path);
	original_file.close(); // closing the original file

	// opening the encrypted file
	std::ifstream encrypted_file(encrypted_filename, std::ios::binary);
	if (!encrypted_file.is_open())
	{
		throw std::runtime_error("File not found");
	}

	// getting the file sizes
	std::filesystem::path original_file_path(file_path);
	size_t original_file_size = std::filesystem::file_size(original_file_path);
	std::filesystem::path encrypted_file_path(encrypted_filename);
	size_t encrypted_file_size = std::filesystem::file_size(encrypted_file_path);


    // The client will try to send the file until it receives a correct CRC from the server or until it reaches the maximum number of attempts
	// The file will be sent in packets and client will only wait for the last packet to be acknowledged by the server
	for (size_t i = 0; i <= NUM_OF_TRIES; i++)
    {
		// packet preparation
        size_t packet_number = 1, total_packets = encrypted_file_size / PACKET_SIZE;
		if (encrypted_file_size % PACKET_SIZE != 0) // if the last packet is not full
        {
            total_packets++;
        }

        while(!encrypted_file.eof())
        {
            char buffer[PACKET_SIZE];
            encrypted_file.read(buffer, PACKET_SIZE);

			std::string buffer_string = std::string(buffer, encrypted_file.gcount());

            // preparing the file request
            Request_t file_request;
            file_request.assign_client_id(client_id);
            file_request.code = SEND_FILE;
            file_request.append_to_payload(std::to_string(encrypted_file_size), CONTENT_LENGTH_SIZE);
            file_request.append_to_payload(std::to_string(original_file_size), ORIGINAL_FILE_LENGTH_SIZE);
            file_request.append_to_payload(std::to_string(packet_number), PACKET_NUMBER_SIZE);
            file_request.append_to_payload(std::to_string(total_packets), TOTAL_PACKETS_SIZE);
            file_request.append_to_payload(filename, FILE_NAME_SIZE);
            file_request.append_to_payload(buffer_string, buffer_string.size());

            // sending the final packet
            if(packet_number == total_packets)
            {
				// getting rid of the encrypted file
				encrypted_file.close();
				std::filesystem::remove(encrypted_filename);

                Response_t response = connection_handler.exchange_messages(file_request);
                if(response.code == FILE_RECEIVED)
                {
                    unsigned long int response_cksum = std::stoul(response.read_from_payload(CKSUM_OFFSET, CKSUM_SIZE));
                    if(cksum == response_cksum)
                    {
						// preparing the CRC correct request
                        Request_t crc_request;
                        crc_request.assign_client_id(client_id);
                        crc_request.code = CORRECT_CRC;
                        crc_request.append_to_payload(filename, FILE_NAME_SIZE);
                        Response_t crc_response = connection_handler.exchange_messages(crc_request);
                        if(crc_response.code == MESSAGE_RECEIVED)
                        {
							std::cout << "File " << filename << " size: " << original_file_size << " bytes backed up successfully" << std::endl;
							return;
                        }
                    }
                    else
                    {
                        Request_t crc_request;
                        crc_request.assign_client_id(client_id);
                        crc_request.code = INCORRECT_CRC;
                        crc_request.append_to_payload(filename, FILE_NAME_SIZE);
                        Response_t crc_response = connection_handler.exchange_messages(crc_request);
                    }
                }
            }
            else
            {
                connection_handler.write_request(file_request);
                packet_number++;
            }
        }
    }

	// If the client failed to send the file after the maximum number of attempts, that is, it didn't return yet, 
	// it will send a request to the server to inform it that the file could not be backed up, and then it will throw an exception
    Request_t crc_request;
    crc_request.assign_client_id(client_id);
    crc_request.code = FOURTH_INCORRECT_CRC;
    crc_request.append_to_payload(filename, FILE_NAME_SIZE);
    Response_t crc_response = connection_handler.exchange_messages(crc_request);
    throw std::runtime_error("incorrect crc for the fourth time - file not backed up");
}

File: server/request_handler.py
import os
import struct
import uuid
from protocol import Response, Size, Offset, Code
from crypto_utils import decrypt_file_with_aes, encrypt_with_rsa, generate_aes_key
from cksum import compute_file_crcg


class RequestHandler:
    """ The class to handle the client requests """
    # Dictionary to store all registered client data
    client_data = {}

    def __init__(self, request):
        self.request = request
        self.response = None

    def handle(self):
        handler = {
            825: self.register_user,
            826: self.exchange_keys,
            827: self.login_user,
            828: self.backup_file,
            900: self.correct_crc,
            901: self.incorrect_crc,
            902: self.fourth_incorrect_crc
        }.get(self.request.opcode)

        handler()
        return self.response


    def register_user(self):
        """ Register the user """
        # Extracting payload
        name = self.request.payload.decode('utf-8')

        # Check if the user is already registered
        if name in RequestHandler.client_data:
            self.response = Response(Code.REGISTRATION_FAILED.value, 0, b'')
        else:
            random_uuid = uuid.uuid4().bytes # Generate a random UUID of size 16 bytes
            RequestHandler.client_data[random_uuid] = {'name': name}

            # Creating the response
            payload = struct.pack(f'!{Size.CLIENT_ID_SIZE.value}s', random_uuid)
            self.response = Response(Code.REGISTRATION_SUCCESSFUL.value, len(payload), payload)


    def exchange_keys(self):
        """ Exchange the AES key """
        # Extracting payload
        rsa_key = self.request.payload[ # Extract the RSA key from the request
                  Offset.PUBLIC_KEY_OFFSET.value: Offset.PUBLIC_KEY_OFFSET.value + Size.PUBLIC_KEY_SIZE.value]

        client_id = self.request.client_id
        RequestHandler.client_data[client_id]['rsa_key'] = rsa_key # Store the RSA key in the client's data for future use
        aes_key = generate_aes_key() # Generate an AES key
        RequestHandler.client_data[client_id]['aes_key'] = aes_key # Storing the AES key in the client's data for future use
        encrypted_aes_key = encrypt_with_rsa(rsa_key, aes_key) # Encrypting the AES key using the client's RSA public key

        # Creating the response
        payload = struct.pack(f'!{Size.CLIENT_ID_SIZE.value}s {len(encrypted_aes_key)}s', client_id, encrypted_aes_key)
        self.response = Response(Code.AES_KEY_EXCHANGE.value, len(payload), payload)


    def login_user(self):
        """ Login the user """
        # Extracting payload
        client_id = self.request.client_id
        # Check if the user isn't registered
        if client_id not in RequestHandler.client_data:
            self.response = Response(Code.LOGIN_FAILED.value, 0, b'')
        else:
            aes_key = generate_aes_key()
            self.client_data[client_id]['aes_key'] = aes_key # Storing the AES key in the client's data for future use
            encrypted_aes_key = encrypt_with_rsa(self.client_data[client_id]['rsa_key'], aes_key)

            # Creating the response
            payload = struct.pack(f'!{Size.CLIENT_ID_SIZE.value}s {len(encrypted_aes_key)}s', client_id, encrypted_aes_key)
            self.response = Response(Code.LOGIN_SUCCESSFUL.value, len(payload), payload)


    def backup_file(self):
        """ Backup the file """
        # Extracting payload
        message_content_size = self.request.payload_size - Offset.MESSAGE_CONTENT_OFFSET.value
        payload_format = f'!{Size.CONTENT_LENGTH_SIZE.value}s {Size.ORIGINAL_FILE_LENGTH_SIZE.value}s {Size.PACKET_NUMBER_SIZE.value}s {Size.TOTAL_PACKETS_SIZE.value}s {Size.FILE_NAME_SIZE.value}s {message_content_size}s'
        content_size, orig_file_size, packet_number, total_packets, file_name, message_content = struct.unpack(payload_format, self.request.payload)
        client_id = self.request.client_id
        client_uuid = uuid.UUID(bytes=client_id)
        aes_key = RequestHandler.client_data[client_id]['aes_key']
        file_name = file_name.rstrip(b'\x00').decode('utf-8') # Decode the file name from bytes to string and unpad it
        packet_number = int(packet_number.decode('utf-8').strip('\x00')) # Convert the packet number to an integer and unpad it
        total_packets = int(total_packets.decode('utf-8').strip('\x00')) # Convert the total packets to an integer and unpad it

        decrypted_file_path = f'backupsvr/{client_uuid}/{file_name}'
        encrypted_file_path = f'backupsvr/{client_uuid}/{file_name}.enc'

        if packet_number == 1:
            # Create the backup directories needed if it doesn't exist
            if not os.path.exists('backupsvr/'):
                os.makedirs('backupsvr/')
            if not os.path.exists(f'backupsvr/{client_uuid}'):
                os.makedirs(f'backupsvr/{client_uuid}')
            if os.path.exists(encrypted_file_path):
                os.remove(encrypted_file_path)

        # Write the message content to the encrypted temporary file
        with open(encrypted_file_path, 'ab') as file:
            file.write(message_content)

        # If this is the last packet, decrypt the encrypted file, read the CRC and file size, and send the response
        if packet_number == total_packets:
            decrypt_file_with_aes(encrypted_file_path, decrypted_file_path, aes_key)
            os.remove(encrypted_file_path) # Remove the encrypted file
            crc_str = compute_file_crc(decrypted_file_path)
            crc = int(crc_str.split('\t')[0])
            file_size = int(crc_str.split('\t')[1])

            # Creating the response
            payload_format = f'!{Size.CLIENT_ID_SIZE.value}s I {Size.FILE_NAME_SIZE.value}s I'
            payload = struct.pack(payload_format, client_id, file_size, file_name.encode('utf-8'), crc)
            self.response = Response(Code.FILE_RECEIVED.value, len(payload), payload)


    def correct_crc(self):
        """ Correct CRC received """
        self.response = Response(Code.MESSAGE_RECEIVED.value, len(self.request.client_id), self.request.client_id)

    def incorrect_crc(self):
        """ Incorrect CRC received """
        os.remove(f'backupsvr/{uuid.UUID(bytes=self.request.client_id)}/{self.request.payload.decode("utf-8")}') # Remove the backed up file with the incorrect CRC

    def fourth_incorrect_crc(self):
        """ Fourth incorrect CRC received """
        os.remove(f'backupsvr/{uuid.UUID(bytes=self.request.client_id)}/{self.request.payload.decode("utf-8")}') # Remove the backed up file with the incorrect CRC
        self.response = Response(Code.MESSAGE_RECEIVED.value, len(self.request.client_id), self.request.client_id)